File: app/src/main/java/com/example/mike/mp3player/client/callbacks/metadata/MyMetadataCallback.java
package com.example.mike.mp3player.client.callbacks.metadata;

import android.media.MediaDescription;
import android.media.MediaMetadata;
import android.os.Handler;
import android.support.v4.media.MediaMetadataCompat;
import android.util.Log;

import com.example.mike.mp3player.client.callbacks.AsyncCallback;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

import javax.inject.Inject;

public class MyMetadataCallback extends AsyncCallback<MediaMetadataCompat> {
    private static final String LOG_TAG = "MY_META_DTA_CLLBK";
    private String currentMediaId = null;
    private Set<MetadataListener> metadataListeners;

    @Inject
    public MyMetadataCallback(Handler handler) {
        super(handler);
        this.metadataListeners = new HashSet<>();
    }

    @Override
    public void processCallback(MediaMetadataCompat mediaMetadataCompat) {

        if (mediaMetadataCompat == null || mediaMetadataCompat.getMediaMetadata() == null) {
            return;
        }
        MediaMetadata mediaMetadata = (MediaMetadata) mediaMetadataCompat.getMediaMetadata();
        if (null == mediaMetadata) {
            return;
        }

        String newMediaId = mediaMetadata.getDescription().getMediaId();
        if (newMediaId != null && !newMediaId.equals(currentMediaId)) {
            this.currentMediaId = newMediaId;
            notifyListeners(mediaMetadataCompat);

        }
    }

    private void notifyListeners(MediaMetadataCompat metadata) {
        //StringBuilder sb = new StringBuilder();
        for (MetadataListener listener : metadataListeners) {
            if (null != listener) {
                listener.onMetadataChanged(metadata);
                //sb.append(listener.getClass());
            }
        }
        // Log.i(LOG_TAG, "hit meta data changed " + ", listeners " + metadataListeners.size() + ", " + sb.toString());
    }


    public synchronized void registerMetaDataListener(MetadataListener listener) {
        Log.i(LOG_TAG, "registerMetaDataListener" + listener.getClass());
        metadataListeners.add(listener);
    }

    public synchronized boolean removeMetaDataListener(MetadataListener listener) {
        return metadataListeners.remove(listener);
    }
}

File: client/src/main/java/com/github/goldy1992/mp3player/client/callbacks/metadata/MyMetadataCallback.kt
package com.github.goldy1992.mp3player.client.callbacks.metadata

import android.media.MediaMetadata
import android.support.v4.media.MediaMetadataCompat
import android.util.Log
import com.github.goldy1992.mp3player.commons.LogTagger
import javax.inject.Inject
import kotlin.collections.HashSet

class MyMetadataCallback

    @Inject
    constructor() : LogTagger {

    private var currentMediaId: String? = null

    private val metadataListeners: MutableSet<MetadataListener> = HashSet()

    fun processCallback(data: MediaMetadataCompat) {
        val mediaMetadata = data.mediaMetadata as MediaMetadata
        val newMediaId = mediaMetadata.description.mediaId
        newMediaId?.let {
            if (newMediaId != currentMediaId) {
                currentMediaId = newMediaId
                notifyListeners(data)
            }
        }
    }

    private fun notifyListeners(metadata: MediaMetadataCompat) {
        for (listener in metadataListeners) {
            listener.onMetadataChanged(metadata)
        }
    }

    fun registerMetaDataListener(listener: MetadataListener) {
        Log.i(logTag(), "registerMetaDataListener" + listener.javaClass)
        metadataListeners.add(listener)
    }

    fun removeMetaDataListener(listener: MetadataListener?): Boolean {
        return metadataListeners.remove(listener)
    }

    override fun logTag(): String {
        return "MY_META_DTA_CLLBK"
    }


}